Add branin plot

MechCoder opened the pull request.

MechCoder commented:
![branin](https://cloud.githubusercontent.com/assets/1867024/14373351/02ab8fd0-fd17-11e5-9704-6c8344768ff8.png)

glouppe commented:
This looks good and convincing! However I am a bit lost about this PR and #50. Arent they both adding an example comparing dummy vs gp?

MechCoder commented:
Yes, but I added this to reproduce the results given here (http://arxiv.org/pdf/1402.0929v3.pdf)

Spearmint seems to achieve a minimum of 0.398 after 200 evaluations using the integrated acquisition function, and we achieve 0.398 by maximizing the likelihood.

MechCoder commented:
@glouppe @betatim Would you be able to merge this?

betatim commented:
Is the conclusion then to stop #50 after merging this one?

----

Can you upload the output of the branin plot function?

betatim commented on examples/plot_branin.py:
```diff
@@ -1,54 +1,59 @@
+best_gp_scores = np.zeros((5, 200))
+n_iterations = range(1, 201)
+
+for random_state in range(5):
```
Is it feasible to do ten rounds or would you have to wait too long? (not super important)

MechCoder replied:
Currently it takes ~25 seconds for every round :/

betatim commented on examples/plot_branin.py:
```diff
@@ -1,54 +1,59 @@
+
+mean_dummy_scores = np.mean(best_dummy_scores, axis=0)
+mean_gp_scores = np.mean(best_gp_scores, axis=0)
+std_dummy_scores = np.std(best_dummy_scores, axis=0)
```
Should we be computing the standard error on the mean? Can someone help with figuring this out?

We show the mean, and are interested in how much the mean can vary, which makes me think `sem`.

MechCoder replied:
Yes, I think

betatim commented on examples/plot_branin_acquisition.py:
```diff
@@ -0,0 +1,54 @@
+from skopt import acquisition
+from skopt import gp_minimize
+from skopt.benchmarks import branin
+
```
blank line

MechCoder replied:
This should there right? Separates the rest of the code from the imports.

betatim commented on examples/plot_branin_acquisition.py:
```diff
@@ -0,0 +1,54 @@
+x1_values = np.linspace(-5, 10, 100)
+x2_values = np.linspace(0, 15, 100)
+x_ax, y_ax = np.meshgrid(x1_values, x2_values)
+vals = np.asarray(list(zip(x_ax.ravel(), y_ax.ravel())))
```
Is `np.c_[x_ax.ravel(), y_ax.ravel()]` truly better or just different syntax?

MechCoder commented:
I have addressed your comments. This is the new graph.

![branin](https://cloud.githubusercontent.com/assets/1867024/14500356/bca0b80a-016f-11e6-8d93-f7c6f38281eb.png)

Also, the other pull request deals with using a sklearn estimator combined with `gp_minimize`. Should I remove the comparison with `dummy_minimize` in this case.

glouppe commented on skopt/__init__.py:
```diff
@@ -1,2 +1,3 @@
+from .gp_opt import acquisition
```
Not sure what is best yet regarding the API, but I would rather not expose `gp_opt.acquisition` at the level of the package. `gp_opt.acquisition` is currently specific to GPs. 

For the example, I would rather do `from skopt.gp_opt import acquisition`

MechCoder replied:
Sure

glouppe commented:
To make things clearer, I would do something like this:
- keep `plot_branin.py` of this PR, but rename to `plot_comparison.py`.
- keep `plot_branin_acquisition.py` of this PR, but rename to `plot_gp_minimize_2d.py`. In this example, we should show the posterior mean of for the objective function, the posterior std for the objective function and the acquisition function (only one of them). Maybe we could have these 3 plots repeated for epoch=1, epoch=N/2 and epoch=N?
- for consistency, I would rename `plot_1d_demo.py` to `plot_gp_minimize_1d.py` and show the same corresponding plots (except that they are now in 1d)
- in #50, rename `plot_random_vs_gp.py` to `plot_hyperparameter_optimization.py` and focus on using gp_minimize as a replacement of GridSearch.

MechCoder commented:
All right, I've addressed the first of your comments. I'll address the remaining 3 in follow up PR's

betatim commented:
That is a nice looking plot and thanks for addressing the comments.
